baseplate/lib/message_queue.py

```diff
@@ -121,6 +159,175 @@ def close(self) -> None:
+        endpoint = config.Endpoint(f"{self.host}:{self.port}")
+        pool = ThriftConnectionPool(
+            endpoint, size=pool_size, timeout=pool_timeout, max_age=pool_conn_max_age
+        )
+        return pool
+
+    def _update_counters(self, outcome: str) -> None:
+        # This request is no longer queued
+        self.remote_queue_put_length.labels(
+            queue_name=self.name,
+        ).dec()
+        # Increment success/failure counters
+        if outcome == "success":
+            metric = self.remote_queue_put_success
+        else:
+            metric = self.remote_queue_put_fail
+        metric.labels(
+            queue_name=self.name,
+        ).inc()
+
+    def _put_success_callback(self, _: Any) -> None:
+        self._update_counters("success")
+
+    def _put_fail_callback(self, greenlet: Any) -> None:
+        self._update_counters("fail")
+        try:
+            greenlet.get()
+        except Exception as e:
+            logging.info("Remote queue `put` failed, exception found: %s", e)
+
+    def get(self, _: Optional[float] = None) -> bytes:
+        raise NotImplementedError  # This queue type is write-only
+
+    def _try_to_put(self, message: bytes, timeout: Optional[float], start_time: float) -> bool:
+        # get a connection from the pool
+        with self.pool.connection() as protocol:
+            client = RemoteMessageQueueService.Client(protocol)
+            try:
+                client.put(message, timeout)
+
+                # record latency
+                self.remote_queue_put_latency.labels(
+                    queue_name=self.name,
+                ).observe(time.perf_counter() - start_time)
+                return True  # Success
+            # If the server responded with a timeout, raise our own timeout to be consistent with the posix queue
+            except ThriftTimedOutError:
+                raise TimedOutError
+
+    def put(self, message: bytes, timeout: Optional[float] = None) -> Any:
+        # increment in-flight counter
+        self.remote_queue_put_length.labels(
+            queue_name=self.name,
+        ).inc()
+        start_time = time.perf_counter()
+
+        greenlet = gevent.spawn(self._try_to_put, message, timeout, start_time)
+        greenlet.link_value(self._put_success_callback)
+        greenlet.link_exception(self._put_fail_callback)
+        return greenlet
+
+
+def create_queue(
+    queue_type: QueueType,
+    queue_full_name: str,
+    max_queue_size: int,
+    max_element_size: int,
+) -> MessageQueue:
+    # The in-memory queue is created on the sidecar, and the main baseplate
+    # application will use a remote queue to interact with it.
+    if queue_type == QueueType.IN_MEMORY:
+        event_queue = InMemoryMessageQueue(max_queue_size)
+
+    else:
+        event_queue = PosixMessageQueue(  # type: ignore
+            queue_full_name,
+            max_messages=max_queue_size,
+            max_message_size=max_element_size,
+        )
```

Comment:
```suggestion
    if queue_type == QueueType.IN_MEMORY:
        log.debug("using in memory queue")
        event_queue = InMemoryMessageQueue(max_queue_size)
    elif queue_type == QueueType.POSIX:
        log.debug("using posix queue")
        event_queue = PosixMessageQueue(  # type: ignore
            queue_full_name,
            max_messages=max_queue_size,
            max_message_size=max_element_size,
        )
    else:
        raise "unknown queue type"
```

its hard to figure out what queue type the sidecar is actually going to use, so adding some logging will be useful. Also, if an incorrect queue type is specified, we should raise vs. silently squelching and using posix.